File: 18.4sum.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> ret;
        if (nums.size() < 4) {
            return ret;
        }
        sort(nums.begin(), nums.end());
        for (int i = 0; i < nums.size() - 3; i++) {
            for (int j = i + 1; j < nums.size() - 2; j++) {
                int t = target - nums[i] - nums[j];
                int k = j + 1, l = nums.size() - 1;
                while (k < l) {
                    if (nums[k] + nums[l] == t) {
                        ret.push_back(vector<int>{nums[i], nums[j], nums[k], nums[l]});
                        while (k < l && nums[k + 1] == nums[k]) {
                            k++;
                        }
                        while (k < l && nums[l - 1] == nums[l]) {
                            l--;
                        }
                        k++;
                        l--;
                    } else if (nums[k] + nums[l] > t) {
                        l--;
                    } else {
                        k++;
                    }
                }
                while (j + 1 < nums.size() && nums[j + 1] == nums[j]) {
                    j++;
                }
            }
            while (i + 1 < nums.size() && nums[i + 1] == nums[i]) {
                i++;
            }
        }
        return ret;
    }
};

File: 18.4sum.py
class Solution:
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        ret = []
        if len(nums) < 4:
            return ret
        nums.sort()
        i = 0
        while i < len(nums) - 3:
            j = i + 1
            while j < len(nums) - 2:
                t = target - nums[i] - nums[j]
                k, l = j + 1,len(nums) - 1
                while k < l:
                    if nums[k] + nums[l] == t:
                        ret.append([nums[i], nums[j], nums[k], nums[l]])
                        while k < l and nums[k + 1] == nums[k]:
                            k += 1
                        while k < l and nums[l - 1] == nums[l]:
                            l -= 1
                        k += 1
                        l -= 1
                    elif nums[k] + nums[l] > t:
                        l -= 1
                    else:
                        k += 1
                while j + 1 < len(nums) and nums[j + 1] == nums[j]:
                    j += 1
                j += 1
            while i + 1 < len(nums) and nums[i + 1] == nums[i]:
                i += 1
            i += 1
        return ret